File: vnes-emulator/src/main/java/vnes/emulator/ui/GUIAdapter.java
package vnes.emulator.ui;
/*
vNES
Copyright © 2006-2013 Open Emulation Project

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import vnes.emulator.NES;
import vnes.emulator.input.InputHandler;
import vnes.emulator.utils.HiResTimer;

/**
 * Adapter class that implements the GUI interface by delegating to components
 * created by a NESUIFactory.
 */
public class GUIAdapter implements GUI {
    private final InputHandler inputHandler;
    private final ScreenView screenView;
    private final HiResTimer timer;

    /**
     * Creates a new GUIAdapter with the specified components.
     * 
     * @param inputHandler The input handler to use
     * @param screenView The screen view to use
     */
    public GUIAdapter(InputHandler inputHandler, ScreenView screenView) {
        this.inputHandler = inputHandler;
        this.screenView = screenView;
        this.timer = new HiResTimer();
    }

    @Override
    public InputHandler getJoy1() {
        return inputHandler;
    }

    @Override
    public InputHandler getJoy2() {
        // Currently only supporting one input handler
        return null;
    }

    @Override
    public ScreenView getScreenView() {
        return screenView;
    }

    @Override
    public HiResTimer getTimer() {
        return timer;
    }

    @Override
    public void imageReady(boolean skipFrame) {
        screenView.imageReady(skipFrame);
    }

    @Override
    public void init(NES nes, boolean showGui) {
        screenView.init();
    }

    @Override
    public void println(String s) {
        System.out.println(s);
    }

    @Override
    public void showErrorMsg(String msg) {
        System.err.println("ERROR: " + msg);
    }

    @Override
    public void showLoadProgress(int percentComplete) {
        // Default implementation does nothing
    }

    /**
     * Cleans up resources used by this GUI adapter.
     */
    public void destroy() {
        if (inputHandler != null) {
            inputHandler.destroy();
        }

        if (screenView != null) {
            screenView.destroy();
        }
    }
}

File: knes-emulator/src/main/kotlin/knes/emulator/ui/GUIAdapter.kt
/*
 *
 *  * Copyright (C) 2025 Artur Skowroński
 *  * This file is part of kNES, a fork of vNES (GPLv3) rewritten in Kotlin.
 *  *
 *  * vNES was originally developed by Brian F. R. (bfirsh) and released under the GPL-3.0 license.
 *  * This project is a reimplementation and extension of that work.
 *  *
 *  * kNES is licensed under the GNU General Public License v3.0.
 *  * See the LICENSE file for more details.
 *
 */

package knes.emulator.ui

import knes.emulator.input.InputHandler
import knes.emulator.utils.HiResTimer

/**
 * Adapter class that implements the GUI interface by delegating to components
 * created by a NESUIFactory.
 */
class GUIAdapter(
    private val inputHandler: InputHandler,
    private val screenView: ScreenView
) : GUI {
    private val timer: HiResTimer = HiResTimer()

    override fun getJoy1(): InputHandler {
        return inputHandler
    }

    override fun getJoy2(): InputHandler? {
        return null
    }

    override fun getTimer(): HiResTimer {
        return timer
    }

    override fun imageReady(skipFrame: Boolean, buffer: IntArray) {
        screenView.imageReady(skipFrame, buffer)
    }

    override fun sendErrorMsg(message: String) {
        System.err.println("ERROR: $message")
    }

    override fun sendDebugMessage(message: String) {}

    override fun destroy() {
        screenView.destroy()
    }
}
